### ISSUE TITLE
[Design] Notification with external apps

### ISSUE THREAD

**Issue by pedroferreira1** [HathorNetwork/hathor-wallet-headless#96]
# Motivation

The wallet and connection objects emit some events when specific actions happen. It would be great for some use cases to capture those events and execute some functions, for example it's important to know when a new transaction arrives for the wallet in order to double check a deposit. If we don't have those event handlers, the use case would need to keep polling the `/tx-history` API to get the latest transactions of the wallet.

# Proposed solution

We currently have the following events:

`HathorWallet`: `state`, `new-tx`, `update-tx`.
`Connection`: `state`, `wallet-update`, `best-block-update`, `wallet-load-partial-update`.

Depending on each use case, it might be interesting to handle any of these events, mainly `new-tx` to execute actions when the wallet receives a new transaction and `state` to know when the wallet has lost connection to the full node.

The range of possibilities to handle those events is huge, someone may want to log some information, add to a SQS, send a request, call a shell script, anything is possible. Because it's a first version, the idea is to have a simple and powerful solution, that can handle all the cases.

We will have a js callback for each event the user wants to handle, configured in the configuration file (`config.js`) like the example below:

```
  eventHandlers: {
    'connection': {
      'best-block-update': (walletId, data) => {
        console.log('Best block update');
        console.log(walletId);
        console.log(data);
      },
    },
    'wallet': {
      'new-tx': (walletId, data) => {
        console.log('New tx');
        console.log(walletId);
        console.log(data);
      },
      'state': (walletId, data) => {
        console.log('State')
        console.log(walletId);
        console.log(data);
      }
    }
  },
```

The handlers will be separated by `connection` and `wallet` events and each event will have its callback with the `walletID` and event data. Each callback can be coded in a separated file and just imported/called in the config file. With this it's possible to execute requests, call shell scripts, log information, ...

The code to add those event handlers in the object is as simple as:

```
  if (config.eventHandlers) {
    const connHandlers = config.eventHandlers.connection || {};
    for (const handlerKey in connHandlers) {
      connection.on(handlerKey, (data) => {
        const cb = connHandlers[handlerKey]
        cb(walletID, data);
      });
    }

    const walletHandlers = config.eventHandlers.wallet || {};
    for (const handlerKey in walletHandlers) {
      wallet.on(handlerKey, (data) => {
        const cb = walletHandlers[handlerKey]
        cb(walletID, data);
      });
    }
  }
```

**Important**

If we decide to go through this solution we must have a better documentation of the possible events and data each event emits.

# Questions

1. Is it enough for the first version to allow only running js code when an event is emitted? Should we add other specific handlers, e.g. running shell script? I think it's enough to run the js callback.
2. Should we separate the event handlers in another file? The `config.js` file is becoming big and complex more than a configuration should be. - I think v1 could keep everything in the config file.

**Comment by msbrogli** [comment]
I like the proposed approach and I feel we should have some builtin handlers, such as `runScriptHandler("/usr/notify.sh")` which will run the cmd passing information using either envvars or the cmdline.

**Comment by msbrogli** [comment]
For me, this design is approved.

**Comment by andreabadesso** [comment]
For me, the design is also approved.

I like the callback idea and the design, but I wouldn't create custom handlers on the headless itself, there are plenty of automation solutions out there (node-red for instance) that can plug on the headless to automate processes

E.g:

Node-red can with a low-code approach, without writing a single line of code, receive an `new-tx` event, check if it is voided and depending on the output addresses, notify other services

@msbrogli

**Comment by pedroferreira1** [comment]
I agree with @andreabadesso. I think we should create the tools and running a shell script is simple using js code. Besides that we don't even know what's the common usage for handling those events. We might create some builtin handlers that are never used and we need to continue giving support to them.

**Comment by msbrogli** [comment]
I just feel that a `runScriptHandler(...)` is always useful for devops. Running a custom scripts seems the most common approach because this is not a library. And, despite how easy it is to call a script, it might not be trivial to handle all the exceptions around it in a safe way to keep the wallet running.

Even solutions as node-red might benefit from it. I'm wondering how node-red would be notified that something happened. Maybe they will need a `runScriptHandler("notify-node-red")`.

**Comment by pedroferreira1** [comment]
> I just feel that a runScriptHandler(...) is always useful for devops. Running a custom scripts seems the most common approach because this is not a library. And, despite how easy it is to call a script, it might not be trivial to handle all the exceptions around it in a safe way to keep the wallet running.

I believe that all use cases we have nowadays would not use a shell script. I think all of them would prefer to create a js method to handle the notification, sending it to a queue, ws message or http endpoint.


I agree about the devops part. I don't think it's a hard/bad feature to have, we can add this.

**Comment by pedroferreira1** [comment]
After discussing this with @msbrogli and @andreabadesso and the main conclusion we all agree is that the most important part of the solution is that it must never handle the events in the main loop of js, thus the first proposed solution in the design is not good.

We have two possible options that can be expanded as we decide to move forward:

1. Create a script runner built in handler as suggested [here](https://github.com/HathorNetwork/hathor-wallet-headless/issues/96#issuecomment-843076700). The main advantage of this solution is that it's as generic as possible because the script can be a python script, shell script or anything the user needs. The downside of it is that we are limited to the packages the container running the headless wallet will have, i.e. if the container does not have python installed, it can't run a python script.
2. Create a handler that sends the events to a queue (we could start with zeromq for instance). This is the way bitcoin usually handles integration but it requires the user to manage a queue service and, from the queue events run its own script.

In both options the idea is to create a generic handler class, in order to be easy enough for anyone to create a new handler and send a PR to us. So we can start with 1 or 2 (or both, if we think it's useful and not hard to implement) and someone may need a HTTP handler, or WS handler, so they can easily construct a new handler and it will work just like our builtin handlers.

It would be great to have a devops view in both options. @luislhl can you share your thoughts about how useful would be both options for devops (e.g. monitoring)?

**Comment by luislhl** [comment]
> It would be great to have a devops view in both options. @luislhl can you share your thoughts about how useful would be both options for devops (e.g. monitoring)?

For monitoring I think I would prefer something that just logs the events for me so I could capture them using some log aggregation tool, because log aggregation is something that a devops will always have already setup in its toolchain.

If those events are logged together with the other hathor-wallet-headless logs (with some info in the logs to identify the type of each log), that would be even better. But with the `script runner` (option 1) I could also easily log everything I need in some file.

I don't see how option 2 would help me for devops, unless I need to integrate with more specific external tools. But maybe a HTTP handler would be more helpful, depending on which external tool I wanted.

But it seems to me that option 2 could be more useful to a lot other use cases that are not devops.

**Comment by luislhl** [comment]
Just one more detail about devops. Maybe we will also want to make hathor-wallet-headless export Prometheus metrics for monitoring, but I'm not yet sure since I didn't do any design on this yet.

For this, some other handler like HTTP can be helpful, because we could create a small service that receives events and transforms them in Prometheus metrics. 

But again, I think devops is not a decisive factor on this decision.
Having this Generic Handler, I can even help by building other handlers, if needed for our devops solutions.

**Comment by pedroferreira1** [comment]
> But it seems to me that option 2 could be more useful to a lot other use cases that are not devops.

I agree with this statement. I think we could go with both options in the first implementation. I feel the hardest work will be to create a generic handler, then it's a small job to create a `script runner` handler or a `queue` handler.

**Comment by giovanecosta** [comment]
>The downside of it is that we are limited to the packages the container running the headless wallet will have, i.e. if the container does not have python installed, it can't run a python script.

Wouldn't WebSocket be a good solution for this problem? I was trying to think from the user PoV: I have a service that is running making requests to a wallet. I probably have an API for my own actions, but maybe i don't. If i want to react to some event from the wallet i would have to put a function in config file that call some endpoint of my application. 

So i would have to make an API to receive those requests or have just a single endpoint that also receives the event as param like a Webhook. Also the server running the wallet, must be able to reach the server running my service.

WebSocket doesn't have any of those problems. I just have to connect and listen. I don't even have to restart wallet server if i want to listen a new event or install packages into container to run my scripts.

**Comment by msbrogli** [comment]
@giovanecosta I guess you have a point. I don't feel websockets are the best to notify external apps but it might be useful for integrations. Maybe it is a missing piece in some of the integrations that are currently using the wallet headless.

**Comment by pedroferreira1** [comment]
After discussing in a meeting with @andreabadesso @msbrogli @giovanecosta @luislhl we've decided to start with a low level design of a concrete case we already have, which is to handle notification for monitoring the headless wallet.

We will create a low level design to execute a command line when a new event arrives in the headless wallet. This low level design will be created when @luislhl starts working in the headless wallet monitoring.

Most important point of the design: we can never handle the events in the js main loop.
